modules/rostests/apitests/shlwapi/IsQSForward.cpp

```diff
@@ -0,0 +1,371 @@
+        }
+        _SEH2_END;
+
+        ok(ret1 == ret2, "cmdID: %ld (%ld vs %ld)\n", cmdID, ret1, ret2);
+        ok((bExcept1 && bExcept2), "cmdID: %ld (%d vs %d)\n", cmdID, bExcept1, bExcept2);
+
+        pGUID = &CGID_Explorer;
+        ret1 = IsQSForward(*pGUID, cCmds, cmds);
+        ret2 = IsQSForwardMockup(pGUID, cCmds, cmds);
+        ok(ret1 == ret2, "cmdID: %ld (%ld vs %ld)\n", cmdID, ret1, ret2);
+
+        pGUID = &IID_IUnknown;
+        ret1 = IsQSForward(*pGUID, cCmds, cmds);
+        ret2 = IsQSForwardMockup(pGUID, cCmds, cmds);
+        ok(ret1 == ret2, "cmdID: %ld (%ld vs %ld)\n", cmdID, ret1, ret2);
+    }
+
+    cCmds = 2;
+    for (cmdID = LOW_VALUE; cmdID <= HIGH_VALUE; ++cmdID)
+    {
+        cmds[0].cmdID = cmdID;
+
+        pGUID = NULL;
+        for (cmdID2 = LOW_VALUE; cmdID2 <= HIGH_VALUE; ++cmdID2)
+        {
+            cmds[1].cmdID = cmdID2;
+            ret1 = IsQSForward(*pGUID, cCmds, cmds);
+            ret2 = IsQSForwardMockup(pGUID, cCmds, cmds);
+            ok(ret1 == ret2, "cmdID: %ld (%ld vs %ld)\n", cmdID, ret1, ret2);
+        }
+
+        pGUID = &CGID_Explorer;
+        for (cmdID2 = LOW_VALUE; cmdID2 <= HIGH_VALUE; ++cmdID2)
+        {
+            cmds[1].cmdID = cmdID2;
+            ret1 = IsQSForward(*pGUID, cCmds, cmds);
+            ret2 = IsQSForwardMockup(pGUID, cCmds, cmds);
+            ok(ret1 == ret2, "cmdID: %ld (%ld vs %ld)\n", cmdID, ret1, ret2);
+        }
+    }
+}
+
+static VOID TEST_MayQSForwardMockup(VOID)
+{
+    CDummyClass dummy;
+    LONG cmdID;
+    OLECMD cmds[1];
+    OLECMDTEXT cmdText;
+    HRESULT ret1, ret2;
+
+    // OLECMDID_PROPERTIES
+    cmds[0].cmdID = cmdID = OLECMDID_PROPERTIES;
+
+    g_bQueryStatus = g_bExec = FALSE;
+    ret1 = MayQSForward(&dummy, 1, CGID_Explorer, _countof(cmds), cmds, &cmdText);
+    ok_int(g_bQueryStatus, FALSE);
+    ok_int(g_bExec, FALSE);
+
+    g_bQueryStatus = g_bExec = FALSE;
+    ret2 = MayQSForwardMockup(&dummy, 1, &CGID_Explorer, _countof(cmds), cmds, &cmdText);
+    ok_int(g_bQueryStatus, FALSE);
+    ok_int(g_bExec, FALSE);
+
+    ok(ret1 == ret2, "cmdID: %ld (%ld vs %ld)\n", cmdID, ret1, ret2);
+
+    // OLECMDID_SHOWFIND
+    cmds[0].cmdID = cmdID = OLECMDID_SHOWFIND;
+
+    g_bQueryStatus = g_bExec = FALSE;
+    ret1 = MayQSForward(&dummy, 1, CGID_Explorer, _countof(cmds), cmds, &cmdText);
+    ok_int(g_bQueryStatus, TRUE);
+    ok_int(g_bExec, FALSE);
+
+    g_bQueryStatus = g_bExec = FALSE;
+    ret2 = MayQSForwardMockup(&dummy, 1, &CGID_Explorer, _countof(cmds), cmds, &cmdText);
+    ok_int(g_bQueryStatus, TRUE);
+    ok_int(g_bExec, FALSE);
+
+    ok(ret1 == ret2, "cmdID: %ld (%ld vs %ld)\n", cmdID, ret1, ret2);
+}
+
+static VOID TEST_MayExecForwardMockup(VOID)
+{
+    CDummyClass dummy;
+    LONG cmdID;
+    HRESULT ret1, ret2;
+
+    // OLECMDID_PROPERTIES
+    cmdID = OLECMDID_PROPERTIES;
+
+    g_bQueryStatus = g_bExec = FALSE;
+    ret1 = MayExecForward(&dummy, 1, CGID_Explorer, cmdID, 0, NULL, NULL);
```

Comment:
Usually, unknown fields (the 2nd one) are set to zero in function calls. Why do you set it to 1 instead? Have you observed some sort of behavioural change on Windows?

Reply:
Yeah, `1` is a positive number. The 2nd parameter should be positive to get successful results.